### PR TITLE
Distance Sensor vl53l1x Enhancements (Added multiple Region of Interest settings)

### PR DESCRIPTION
**Pull request by dirksavage88**
Increased timing budget, added multiple ROI zones.

Max Range with 200ms timing budget is now ~3.8m.
Added 15 zones (Region of interest) and scan through zones to increase resolution of range data.

Originally pull request #16695

### CODE REVIEW

**Comment by dirksavage88** [comment]
Will upload flight log if required @dagar

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 1]
```diff
@@ -479,6 +531,24 @@ int8_t VL53L1X::VL53L1X_StopRanging()
 	return status;
 }
 
+int8_t VL53L1X::VL53L1X_SetROI(uint16_t x, uint16_t y)
+{
+	int8_t status = 0;
+
+	status = VL53L1_WrByte(ROI_CONFIG__USER_ROI_REQUESTED_GLOBAL_XY_SIZE,
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:538]:
```suggestion
	int status = VL53L1_WrByte(ROI_CONFIG__USER_ROI_REQUESTED_GLOBAL_XY_SIZE,
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 2]
```diff
@@ -221,30 +260,43 @@ int VL53L1X::probe()
...
 
-	ScheduleDelayed(VL53L1X_SAMPLE_RATE);
+	ScheduleDelayed(VL53L1X_DELAY);
+
+	// zone modulus increment
+
+	zone_index = (zone_index + 1) % zone_limit;
+
+	// Set the ROI center based on zone incrementation
+
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:276]:
```suggestion
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 3]
```diff
@@ -221,30 +260,43 @@ int VL53L1X::probe()
...
 
 	if (dataReady == 1) {
 		collect();
 	}
 
-	ScheduleDelayed(VL53L1X_SAMPLE_RATE);
+	ScheduleDelayed(VL53L1X_DELAY);
+
+	// zone modulus increment
+
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:272]:
```suggestion
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 4]
```diff
@@ -1,6 +1,6 @@
 /****************************************************************************
  *
- *   Copyright (c) 2020 PX4 Development Team. All rights reserved.
+ *   Copyright (c) 2021 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:3]:
```suggestion
 *   Copyright (c) 2020-2021 PX4 Development Team. All rights reserved.
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp [hunk 5]
```diff
@@ -479,6 +531,24 @@ int8_t VL53L1X::VL53L1X_StopRanging()
...
+			       (y - 1) << 4 | (x - 1));     /* set ROI size x and y */
+
+	return status;
+}
+
+int8_t VL53L1X::VL53L1X_SetROICenter(uint8_t zone)
+{
+	int8_t status = 0;
+
+	status = VL53L1_WrByte(VL53L1_ROI_CONFIG__MODE_ROI_CENTRE_SPAD, zone);    /* Set ROI spad center */
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.cpp:548]:
```suggestion
	int status = VL53L1_WrByte(VL53L1_ROI_CONFIG__MODE_ROI_CENTRE_SPAD, zone);    /* Set ROI spad center */
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.hpp [hunk 6]
```diff
@@ -117,6 +117,15 @@ class VL53L1X : public device::I2C, public I2CSPIDriver<VL53L1X>
 	 */
 	void RunImpl();
 
+	// Distance mode member variable
+	uint16_t distance_mode;
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.hpp:121]:
```suggestion
	uint16_t _distance_mode{VL53L1X_LONG_RANGE};
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.hpp [hunk 7]
```diff
@@ -117,6 +117,15 @@ class VL53L1X : public device::I2C, public I2CSPIDriver<VL53L1X>
...
+
+	// Zone index member variable
+	uint8_t zone_index;
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.hpp:124]:
```suggestion
	uint8_t _zone_index{0};
```

File: src/drivers/distance_sensor/vl53l1x/vl53l1x.hpp [hunk 8]
```diff
@@ -117,6 +117,15 @@ class VL53L1X : public device::I2C, public I2CSPIDriver<VL53L1X>
...
+
+	// Zone limit member variable
+	uint8_t zone_limit;
```
Comment (dagar) [inline src/drivers/distance_sensor/vl53l1x/vl53l1x.hpp:127]:
```suggestion
	uint8_t _zone_limit{0};
```

As a habit it's safer to initialize everything.

**Comment by dagar** [comment]
> Will upload flight log if required @dagar

If you have one it couldn't hurt to take a look. The main thing is to sanity test it for real (in flight, etc) on actual embedded hardware.

**Comment by dirksavage88** [comment]
> > Will upload flight log if required @dagar
> 
> If you have one it couldn't hurt to take a look. The main thing is to sanity test it for real (in flight, etc) on actual embedded hardware.

https://review.px4.io/plot_app?log=4cb45437-2d32-45cc-bbad-763577ed89a2
Screenshot from 2021-08-18 19-55-27

**Comment by dagar** [comment]
I've merged into current master to resolve the minor conflict in your branch. This should be good to go once it makes it through the CI system. Thanks @dirksavage88.

**Review by dagar — APPROVED** [review]

**Pull request merged**